## Libraries imports

In [ ]:
pip install tabula-py

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 81.9 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import numpy as np
import tabula

## Data imports

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/Statapp/data_02.csv")

<ipython-input-6-7ad2a7d11041>:1: DtypeWarning: Columns (6129) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("/content/drive/MyDrive/Statapp/data_02.csv")


In [ ]:
data.shape

(42233, 6130)

In [ ]:
data.head()

HHIDPN  R1MSTAT  R1MPART  R2MSTAT  R2MPART  R3MSTAT  R3MPART  R4MSTAT  \
0      1010      5.0      0.0      5.0      0.0      NaN      NaN      NaN   
1      2010      7.0      0.0      7.0      0.0      7.0      0.0      7.0   
2      3010      1.0      0.0      1.0      0.0      1.0      0.0      1.0   
3      3020      1.0      0.0      1.0      0.0      1.0      0.0      1.0   
4  10001010      8.0      0.0      8.0      0.0      8.0      0.0      8.0   

   R4MPART  R5MSTAT  ...  genetic_4_PP_COGENT17  genetic_4_SBP_COGENT17  \
0      NaN      NaN  ...                    NaN                     NaN   
1      0.0      7.0  ...                    NaN                     NaN   
2      0.0      1.0  ...               -0.58662                -0.66936   
3      0.0      1.0  ...               -0.06204                -0.70205   
4      0.0      8.0  ...                    NaN                     NaN   

   genetic_4_EGFR_CKDGEN19  genetic_4_EGFRTE_CKDGEN19  \
0                      NaN                        NaN   
1                      NaN                        NaN   
2                  1.04423                    1.02390   
3                 -1.06050                   -1.08282   
4                      NaN                        NaN   

   genetic_4_EA3_W23_SSGAC18  genetic_4_HBA1CAA_MAGIC17  \
0                        NaN                        NaN   
1                        NaN                        NaN   
2                    1.53372                    0.47688   
3                    0.85439                   -0.62488   
4                        NaN                        NaN   

   genetic_4_HBA1CEA_MAGIC17  genetic_4_GCOG2_CHARGE18  genetic_VERSION  \
0                        NaN                       NaN              NaN   
1                        NaN                       NaN              NaN   
2                   -1.28083                  -1.51568              4.3   
3                   -1.29031                   0.40118              4.3   
4                        NaN                       NaN              NaN   

   genetic_Section_A_or_E  
0                     NaN  
1                     NaN  
2                       E  
3                       E  
4                     NaN  

[5 rows x 6130 columns]

In [ ]:
# Only if necessary. (It appears to ALWAYS be necessary :-( )

# convert PDF into CSV
tabula.convert_into("/content/drive/MyDrive/Statapp/Documentation_données_socio_économique.pdf", "Documentation_données_socio_économique.csv", output_format="csv", pages='all')

May 08, 2023 6:02:10 PM org.apache.pdfbox.pdmodel.font.FileSystemFontProvider <init>
May 08, 2023 6:02:10 PM org.apache.pdfbox.pdmodel.font.FileSystemFontProvider <init>
May 08, 2023 6:02:10 PM org.apache.pdfbox.pdmodel.font.PDType1Font <init>
May 08, 2023 6:02:10 PM org.apache.pdfbox.pdmodel.font.PDType1Font <init>
May 08, 2023 6:06:02 PM org.apache.pdfbox.pdmodel.font.PDSimpleFont toUnicode



In [ ]:
columns_types = pd.read_csv("Documentation_données_socio_économique.csv", skiprows=299)

<ipython-input-10-90cce48ecf6b>:1: DtypeWarning: Columns (6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42) have mixed types. Specify dtype option on import or set low_memory=False.
  columns_types = pd.read_csv("Documentation_données_socio_économique.csv", skiprows=299)


In [ ]:
columns_types.sample(5)

Household Identifier . . . .                                .  \
19330                           10                        R10GRPPOS   
4251                         2002:                              NaN   
29847                        xH042  MONEY OWED ON 2ND PROPERTY LOAN   
10602                     S9HIBPEF                            10953   
33400        501: 5 in HH/1 kids |                              160   

                                                     ..1    ..2  ..3   ..4  \
19330  R10GRPPOS:W10 R position during grip strength ...  Categ  NaN   NaN   
4251                                                 NaN    NaN  NaN   NaN   
29847                                                NaN    NaN  NaN   NaN   
10602                                               0.07   0.90  0.0  14.0   
33400                                                131    197  138   NaN   

       ..5  ..6  ..7  ..8  ... ..31 ..32 ..33 ..34 ..35 ..36 ..37 ..38 ..39  \
19330  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
4251   NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
29847  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
10602  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
33400  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

      . 60  
19330  NaN  
4251   NaN  
29847  NaN  
10602  NaN  
33400  NaN  

[5 rows x 43 columns]

## Getting types of columns

In [ ]:
columns_types = columns_types.rename(columns={"Household Identifier . . . .": "Wave", ".":"Name", "..2":"Type1", "..3": "Type2"})

columns_types = columns_types[["Name", "Type1", "Type2"]]
columns_types["Type1"] = columns_types["Type1"].fillna(columns_types["Type2"])
columns_types = columns_types.drop(columns=["Type2"]).rename(columns={"Type1": "Type"})

columns_types = columns_types.loc[columns_types["Type"].isin(["Categ", "Cont", "Char"])]

columns_types = columns_types.reset_index(drop=True)

In [ ]:
columns_types.shape

(14447, 2)

In [ ]:
columns_types.head()

Name  Type
0    HHIDPN  Cont
1      HHID  Char
2        PN  Char
3  RAHHIDPN  Char
4    H1HHID  Cont

We now assert that each column types has been recorded. 

In [ ]:
columns_not_recorded = []
for col in data.columns:
  if "genetic" not in col and col not in columns_types["Name"].values:
    columns_not_recorded.append(col)

print(f"Among the {data.shape[1]} columns in our data, {len(columns_not_recorded)} are of unknown type.")

Among the 6130 columns in our data, 267 are of unknown type.


Still 267 columns are to handle. Let us see if we could say something about their type, in view of the number of different values they take.

In [ ]:
uniques_number = {}
for col in columns_not_recorded:
  x = data[col].nunique()
  if x not in uniques_number.keys():
    uniques_number[x] = 1
  else:
    uniques_number[x] += 1
uniques_number = dict(sorted(uniques_number.items()))
for k, v in uniques_number.items():
  print(f"{v} column(s) have exactly {k} values." )

25 column(s) have exactly 2 values.
14 column(s) have exactly 3 values.
41 column(s) have exactly 4 values.
16 column(s) have exactly 5 values.
71 column(s) have exactly 6 values.
15 column(s) have exactly 7 values.
1 column(s) have exactly 8 values.
1 column(s) have exactly 9 values.
1 column(s) have exactly 11 values.
1 column(s) have exactly 13 values.
1 column(s) have exactly 14 values.
1 column(s) have exactly 17 values.
1 column(s) have exactly 18 values.
1 column(s) have exactly 19 values.
1 column(s) have exactly 20 values.
1 column(s) have exactly 21 values.
1 column(s) have exactly 23 values.
1 column(s) have exactly 33 values.
2 column(s) have exactly 36 values.
1 column(s) have exactly 38 values.
1 column(s) have exactly 43 values.
1 column(s) have exactly 80 values.
1 column(s) have exactly 89 values.
1 column(s) have exactly 93 values.
1 column(s) have exactly 104 values.
1 column(s) have exactly 105 values.
1 column(s) have exactly 106 values.
1 column(s) have exactly 11

By checking the documentation, we see that columns which present more thant 17 unique values are continuous and the ones with less than 7 values are categorical. Let save this.

The doubt remains for columns which contain between 8 and 16 unique values. Let us explore them by hand.

In [ ]:
for col in columns_not_recorded:
  if data[col].nunique() <= 17 and data[col].nunique() >= 8:
    print(col, data[col].unique())

REXITWV [ 3.  6. 12. 13. nan  5. 11.  9.  2. 10.  7.  4.  8. 14.]
REXITYR [1996. 2002. 2014. 2016.   nan 2000. 2012. 2008. 1994. 2010. 2004. 1998.
 2006. 2018. 1995.]
R9DRINKN [nan  1.  0.  3.  2.  5.  4. 10.  6. 12.  7.  8. 15. 11. 16. 17.  9. 14.]
R8RETDTWV [nan  8.  7.  6.  1.  5.  3.  4.  2.]
R9RETDTWV [nan  8.  7.  6.  1.  5.  9.  4.  3.  2.]
R1WORKLM [ 50.  nan  30.   0. 100.  20.  40.  80.  60.  90.  10.  70.]


After checking documentation:
- REXITWV, R8RETDTWV, R9RETDTWV are categorical variables
- REXITYR, R9DRINKN, R1WORKLM are continuous variables

In [ ]:
other_columns = {"Name" : ["REXITWV", "R8RETDTWV", "R9RETDTWV", "REXITYR", "R9DRINKN", "R1WORKLM"],
                 "Type" : ["Categ", "Categ", "Categ", "Cont", "Cont", "Cont"]}
for col in columns_not_recorded:
  number_of_values = data[col].nunique()

  if number_of_values >= 17:
    other_columns["Name"].append(col)
    other_columns["Type"].append("Cont")

  elif number_of_values <= 7:
    other_columns["Name"].append(col)
    other_columns["Type"].append("Categ")

columns_types = pd.concat([columns_types, pd.DataFrame(other_columns)]).reset_index(drop=True)

In [ ]:
print(columns_types.shape)
columns_types.head()

(14715, 2)


Name  Type
0    HHIDPN  Cont
1      HHID  Char
2        PN  Char
3  RAHHIDPN  Char
4    H1HHID  Cont

## Dropping health-related variables and saving Global Health Index

The global health index we wish to predict has been created based on health-related variables. To ensure we do not use these variables to predict the index, we shall delete them.

According to the documentation, the health-related variables go from R1SHLT to S14PMWAISTC.


In [ ]:
i_R1SHLT = columns_types.loc[columns_types["Name"] == "R1SHLT"].index[0]
i_S14PMWAISTC = columns_types.loc[columns_types["Name"] == "S14PMWAISTC"].index[0]

health_variables = columns_types.iloc[i_R1SHLT:i_S14PMWAISTC+1, 0].values
health_variables = [var for var in health_variables if var in data.columns]

In [ ]:
print(f"Before deletion of health-related variable, database is of shape {data.shape}.")
data.info(memory_usage="deep")

Before deletion of health-related variable, database is of shape (42233, 6130).
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42233 entries, 0 to 42232
Columns: 6130 entries, HHIDPN to genetic_Section_A_or_E
dtypes: float64(6018), int64(111), object(1)
memory usage: 1.9 GB


In [ ]:
data = data.drop(columns = health_variables)

In [ ]:
print(f"After deletion of health-related variable, database is of shape {data.shape}.")
data.info(memory_usage="deep")

After deletion of health-related variable, database is of shape (42233, 4147).
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42233 entries, 0 to 42232
Columns: 4147 entries, HHIDPN to genetic_Section_A_or_E
dtypes: float64(4047), int64(99), object(1)
memory usage: 1.3 GB


We now save the index.

In [ ]:
data_GHI = pd.read_csv("/content/drive/MyDrive/Statapp/data_tSNE_GHI.csv")
data_GHI.columns = ['HHIDPN', 'GHI1', 'GHI2', 'GHI3', 'GHI4', 'GHI5', 'GHI6', 'GHI7', 'GHI8', 'GHI9', 'GHI10', 'GHI11', 'GHI12', 'GHI13', 'GHI14']
data_GHI.head()

HHIDPN       GHI1       GHI2       GHI3       GHI4       GHI5       GHI6  \
0      1010  46.338680  68.969490        NaN        NaN        NaN        NaN   
1      2010  64.833760 -13.297101  27.734835  43.220688  44.066593        NaN   
2      3010  22.931510 -66.686300 -29.234972  13.921358 -91.249150 -44.982487   
3      3020  87.220850  26.626232  -3.540929 -35.344840   5.839192  15.979321   
4  10001010  16.348322  51.236960 -34.807380  16.018581 -20.836510 -49.169600   

        GHI7       GHI8        GHI9      GHI10      GHI11      GHI12  \
0        NaN        NaN         NaN        NaN        NaN        NaN   
1        NaN        NaN         NaN        NaN        NaN        NaN   
2 -73.210106 -93.987110  -26.113197 -49.452682 -13.593910        NaN   
3  87.244750   7.659104    7.514696  11.224237   6.606269  85.520730   
4 -41.571840 -27.421490  104.275900 -33.672630 -10.305906 -51.743675   

       GHI13  GHI14  
0        NaN    NaN  
1        NaN    NaN  
2        NaN    NaN  
3        NaN    NaN  
4 -60.904327    NaN

In [ ]:
print(data.shape, data_GHI.shape)

(42233, 4147) (42233, 15)


We see that one individual is missing from data_GHI. We will delete him from the original dataset.

In [ ]:
for hhidpn in data["HHIDPN"].values:
  if hhidpn not in data_GHI["HHIDPN"].values:
    print(hhidpn)

In [ ]:
for hhidpn in data_GHI["HHIDPN"].values:
  if hhidpn not in data["HHIDPN"].values:
    print(hhidpn)

In [ ]:
data = data[data["HHIDPN"] != 32570030]

In [ ]:
data = data.merge(data_GHI, on="HHIDPN")
data.head()

HHIDPN  R1MSTAT  R1MPART  R2MSTAT  R2MPART  R3MSTAT  R3MPART  R4MSTAT  \
0      1010      5.0      0.0      5.0      0.0      NaN      NaN      NaN   
1      2010      7.0      0.0      7.0      0.0      7.0      0.0      7.0   
2      3010      1.0      0.0      1.0      0.0      1.0      0.0      1.0   
3      3020      1.0      0.0      1.0      0.0      1.0      0.0      1.0   
4  10001010      8.0      0.0      8.0      0.0      8.0      0.0      8.0   

   R4MPART  R5MSTAT  ...       GHI5       GHI6       GHI7       GHI8  \
0      NaN      NaN  ...        NaN        NaN        NaN        NaN   
1      0.0      7.0  ...  44.066593        NaN        NaN        NaN   
2      0.0      1.0  ... -91.249150 -44.982487 -73.210106 -93.987110   
3      0.0      1.0  ...   5.839192  15.979321  87.244750   7.659104   
4      0.0      8.0  ... -20.836510 -49.169600 -41.571840 -27.421490   

         GHI9      GHI10      GHI11      GHI12      GHI13  GHI14  
0         NaN        NaN        NaN        NaN        NaN    NaN  
1         NaN        NaN        NaN        NaN        NaN    NaN  
2  -26.113197 -49.452682 -13.593910        NaN        NaN    NaN  
3    7.514696  11.224237   6.606269  85.520730        NaN    NaN  
4  104.275900 -33.672630 -10.305906 -51.743675 -60.904327    NaN  

[5 rows x 4161 columns]

In [ ]:
columns_types = pd.concat([columns_types, pd.DataFrame({"Name" : [f"GHI{wave}" for wave in range(1, 15)]+genetic_columns, "Type" : ["Cont"]*99})], ignore_index=True)
columns_types.iloc[14813, 1] = "Char" # We change the type of genetic_Section_A_or_E to 'Char'

In [ ]:
columns_types.to_csv("data_03_columns_types.csv", index=False)

In [ ]:
data.to_csv("data_03.csv", index=False)

## Creating dummy variables

As many of our variables are categorical, performing direct imputation would lead to meaningless results: a variable 'example' that take [a, b, c, d, e] as values may be imputed 2.25 or any other unwanted numeric value. 

To deal with this problem, we will convert each categorical variable into a list of dummy variables. In our previous exemple, we would replace the sole variable that take [a, b, c, d, e] as values by five variables 'example_a', 'example_b', 'example_c', 'example_d', 'example_e' that takes 0 or 1 as values. In this paradigm, having an imputed value of example_e = 0.87 could be seen as the probabilty for 'example' to take value 'e'.

In [ ]:
categorical_variables = columns_types.loc[columns_types["Type"] == "Categ"]["Name"].values
categorical_variables = [col for col in categorical_variables if (col in data.columns and data[col].nunique() != 2)]

In [ ]:
len(categorical_variables)

1462

In [ ]:
data_dummies = pd.get_dummies(data, columns = categorical_variables)
print(data.shape, data_dummies.shape)
data_dummies.head()

(42232, 4161) (42232, 11429)


HHIDPN  R1MPART  R2MPART  R3MPART  R4MPART  R5MPART  R6MPART  R7MPART  \
0      1010      0.0      0.0      NaN      NaN      NaN      NaN      NaN   
1      2010      0.0      0.0      0.0      0.0      0.0      NaN      NaN   
2      3010      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3      3020      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
4  10001010      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

   R8MPART  R9MPART  ...  R13IADL5H_2  R13IADL5H_3  R13IADL5H_4  R13IADL5H_5  \
0      NaN      NaN  ...            0            0            0            0   
1      NaN      NaN  ...            0            0            0            0   
2      0.0      0.0  ...            0            0            0            0   
3      0.0      0.0  ...            0            0            0            0   
4      0.0      0.0  ...            0            0            0            0   

   R14IADL5H_0  R14IADL5H_1  R14IADL5H_2  R14IADL5H_3  R14IADL5H_4  \
0            1            0            0            0            0   
1            1            0            0            0            0   
2            1            0            0            0            0   
3            1            0            0            0            0   
4            1            0            0            0            0   

   R14IADL5H_5  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 11429 columns]

In [ ]:
for col in categorical_variables:
  data_dummies.loc[data[col].isna(), data_dummies.columns.str.startswith(col)] = np.nan
data_dummies.head()

HHIDPN  R1MPART  R2MPART  R3MPART  R4MPART  R5MPART  R6MPART  R7MPART  \
0      1010      0.0      0.0      NaN      NaN      NaN      NaN      NaN   
1      2010      0.0      0.0      0.0      0.0      0.0      NaN      NaN   
2      3010      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
3      3020      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
4  10001010      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

   R8MPART  R9MPART  ...  R13IADL5H_2  R13IADL5H_3  R13IADL5H_4  R13IADL5H_5  \
0      NaN      NaN  ...          0.0          0.0          0.0          0.0   
1      NaN      NaN  ...          0.0          0.0          0.0          0.0   
2      0.0      0.0  ...          0.0          0.0          0.0          0.0   
3      0.0      0.0  ...          0.0          0.0          0.0          0.0   
4      0.0      0.0  ...          0.0          0.0          0.0          0.0   

   R14IADL5H_0  R14IADL5H_1  R14IADL5H_2  R14IADL5H_3  R14IADL5H_4  \
0          1.0          0.0          0.0          0.0          0.0   
1          1.0          0.0          0.0          0.0          0.0   
2          1.0          0.0          0.0          0.0          0.0   
3          1.0          0.0          0.0          0.0          0.0   
4          1.0          0.0          0.0          0.0          0.0   

   R14IADL5H_5  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  

[5 rows x 11429 columns]

## Export data

In [ ]:
data_dummies.to_csv("data_04.csv", index=False)